Mathlib/Data/Nat/Digits/DigitsFin.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2025 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import Mathlib.Algebra.BigOperators.Fin
+import Mathlib.Data.Nat.Digits.Lemmas
+
+/-!
+# First digits of a natural number
+
+This file provides a basic API for extracting the first `l` digits of a natural number `n` in a
+given base `b` as a function `Fin l → ℕ`.
+
+This is useful for computations with natural numbers of bounded length. For example, we provide a
+proof that the natural numbers `< b ^ l` are in bijection with the functions
+`Fin l → Finset.range b`, see `Nat.digitsFin_bijOn`. See also `Nat.sum_digits_sum_eq` for an
+application.
+
+## Main definitions and results
+
+* `Nat.digitsFin`: the function that return the first `l` digits of `n` in some base `b`.
+  Unlike `Nat.digits`, the last digits can be equal to `0` (this happens if `n < b ^ l`).
+* `Nat.digitsFin_eq_append`: `b.digitsFin l n` is the concatenation of the list of digits of `n`
+  in base `b` and some `0`'s.
+* `Nat.digitsFin_bijOn`: the map `Nat.digitsFin l b` gives a bijection between the set of
+  natural numbers `< b ^ l` and the set of functions `Fin l → Finset.range b`.
+* `Nat.sum_digits_sum_eq`: formula for the sum of the sum of digits in base `b` of all natural
+  integers `< b ^ l`.
+-/
+
+namespace Nat
+
+open List
+
+variable (b l n : ℕ)
+
+theorem ofDigits_ofFn (f : Fin l → ℕ) :
+    ofDigits b (ofFn f) = ∑ i, f i * b ^ i.val := by
+  induction l with
+  | zero => simp
+  | succ n hn =>
+      rw [ofFn_succ, ofDigits_cons, Fin.sum_univ_succ, hn, Finset.mul_sum, Fin.val_zero, pow_zero,
+        mul_one]
+      simp_rw [Nat.mul_left_comm b, ← pow_succ', Fin.val_succ]
+
+/--
+The first `l` digits, in little-endian order, of a natural number `n` in a specified base `b`.
+Unlike `Nat.digits`, the last digits can be equal to `0` if `n` is too small.
+-/
+def digitsFin : ℕ → Fin l → ℕ := fun x i ↦ (b.digits x)[i]?.getD 0
```

Comment:
This isn't the signature I'd expect this function to have; I would have expected `digitsFin b n : ℕ → Fin b`, probably with `NeZero b` or similar.

Reply:
I agree with Eric here, summing from Nat is usually easier, particularly for summation, and typically we use lemmas like [Fin.sum_univ_eq_sum_range](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Fintype/BigOperators.html#Fin.sum_univ_eq_sum_range) to convert to a sum over Nat anyway.

Reply:
Ok, I will change the type of `digitsFin` to  `ℕ → ℕ → ℕ`.

Reply:
or
```lean
/-- The `i`th digit of `n` in base `b`. -/
def Nat.finDigit (b : Nat) [NeZero b] (n : Nat) (i : Nat) : Fin b :=
  ⟨(n / (b ^ i)) % b, Nat.mod_lt _ <| Nat.pos_of_neZero b⟩
```

Reply:
I am still not sure about having the function returning a `Fin b`.